packages/kit/src/exports/vite/index.js

```diff
@@ -529,71 +562,81 @@ function kit({ svelte_config }) {
 		 */
 		writeBundle: {
 			sequential: true,
-			async handler(_options, bundle) {
-				if (vite_config.build.ssr) return;
-
-				const guard = module_guard(this, {
-					cwd: vite.normalizePath(process.cwd()),
-					lib: vite.normalizePath(kit.files.lib)
-				});
-
-				manifest_data.nodes.forEach((_node, i) => {
-					const id = vite.normalizePath(path.resolve(kit.outDir, `generated/nodes/${i}.js`));
-
-					guard.check(id);
-				});
+			async handler(_options) {
+				if (secondary_build) return; // only run this once
+				secondary_build = true;
 
 				const verbose = vite_config.logLevel === 'info';
-				log = logger({
-					verbose
-				});
-
-				const { assets, chunks } = collect_output(bundle);
-				log.info(
-					`Client build completed. Wrote ${chunks.length} chunks and ${assets.length} assets`
-				);
+				const log = logger({ verbose });
 
-				log.info('Building server');
+				/** @type {import('vite').Manifest} */
+				const server_manifest = JSON.parse(read(`${out}/server/${vite_config.build.manifest}`));
 
-				const options = {
-					config: svelte_config,
-					vite_config,
-					vite_config_env,
+				/** @type {import('types').BuildData} */
+				const build_data = {
+					app_dir: kit.appDir,
+					app_path: `${kit.paths.base.slice(1)}${kit.paths.base ? '/' : ''}${kit.appDir}`,
 					manifest_data,
-					output_dir: out
+					service_worker: !!service_worker_entry_file ? 'service-worker.js' : null, // TODO make file configurable?
+					client_entry: null,
+					server_manifest
 				};
 
-				/** @type {import('vite').Manifest} */
-				const vite_manifest = JSON.parse(
-					fs.readFileSync(`${out}/client/${vite_config.build.manifest}`, 'utf-8')
+				const manifest_path = `${out}/server/manifest-full.js`;
+				fs.writeFileSync(
+					manifest_path,
+					`export const manifest = ${generate_manifest({
+						build_data,
+						relative_path: '.',
+						routes: manifest_data.routes
+					})};\n`
 				);
 
-				const client = {
-					assets,
-					chunks,
-					entry: find_deps(
-						vite_manifest,
-						posixify(path.relative('.', `${runtime_directory}/client/start.js`)),
-						false
-					),
-					vite_manifest
-				};
+				// first, build server nodes without the client manifest so we can analyse it
```

Comment:
There's two ways you could so that, and both suck:

* you spin up the Vite dev server and load the routes through `ssrLoadModule`, which involves lots of overhead and wouldn't be 100% reliable (e.g. if something depends on `dev` being `false` for whatever reason)
* you statically analyse the code, which involves lots of overhead, and will fail if a) we one day need to do more than check for the existence of a particular export, or b) the code has syntax that needs to be transformed (e.g. a futuristic TypeScript version that SvelteKit hasn't yet upgraded to)